packages/mui-utils/src/getReactNodeRef/getReactNodeRef.ts

```diff
@@ -8,13 +8,9 @@ import * as React from 'react';
  * @returns React.Ref<any> | null
  */
 export default function getReactNodeRef(element: React.ReactNode): React.Ref<any> | null {
-  if (!element || !React.isValidElement(element)) {
-    return null;
-  }
-
   // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in older versions
-  return (element.props as any).propertyIsEnumerable('ref')
-    ? (element.props as any).ref
+  return (element as any)?.props?.propertyIsEnumerable('ref')
```

Comment:
I'm unsure if type if element should be `React.ReactNode` or `React.ReactElement`, currently this util is used by `Portal` and portal supports multiple children 
https://github.com/mui/material-ui/blob/0137be75949f5a2645ce13272ca8d952dbbc0c6e/packages/mui-base/src/Portal/Portal.tsx#L5

But if type of `element` remains as `React.ReactNode` we need to type case `element` to `any` to make `element.props.propertyIsEnumerable('ref')` work.

Where as if `element` type is `React.ReactElement`, type casting to `any` is not necessary

Reply:
Taking into account that:
- The recommended way to use this helper seems to check with `React.isValidElement` first (as per https://github.com/mui/material-ui/pull/43022#discussion_r1687208774)
- `getReactNodeRef`'s behaviour goes like this:
  - returns the ref if the param is an element with a ref
  - returns null for elements without a ref and fragments
  - returns undefined for arrays.
  - throws an error when passing "non-object" children types (null, undefined, booleans, string, etc.)

Maybe it should be named `getReactElementRef` instead of `getReactNodeRef`? And update the param type to be `React.ReactElement` instead of `React.ReactNode`.

Here's a playground showcasing the differences between `getReactElementRef` and `getReactNodeRef`. Note how the former forces consumers to check with `React.isValidElement` before using it, which I think it's a better DX and avoids unexpected errors.

https://codesandbox.io/p/sandbox/getreactnoderef-zjcv2t?file=/src/App.tsx:73,21

Let me know if this makes sense or if I missed something.

cc @DiegoAndai since you worked on this helper, too.

Reply:
The logic behind leaving the `isValidElement` check inside the util is that we were doing the check on most places in which we used the util so it seemed better to incorporate it into it (https://github.com/mui/material-ui/pull/43132#discussion_r1699451479). I'm not against moving the check out, but I would like to understand what's the benefit of doing so. How is forcing consumers to check `isValidElement` better DX?

Reply:
> if type if element should be React.ReactNode or React.ReactElement

+1 for React.ReactElement

> How is forcing consumers to check isValidElement better DX?

@DiegoAndai I tried to demonstrate it in https://github.com/mui/material-ui/pull/43022#discussion_r1687208774. Most of the time, there is already a warning in place to flag when using the API silencing the error up in the the stack makes it easy for developers to notice the problem.

Reply:
Ok, I'm on board with @oliviertassinari and @aarongarciah's proposed changes, lets go with the approach in https://github.com/mui/material-ui/pull/43022#discussion_r1715199139 @sai6855.

Reply:
Sounds great 👍. 

Before we move forward, I think to sync with https://github.com/mui/base-ui/pull/605#discussion_r1759927875 so we move in the same direction.

Reply:
:+1: We're checking React version and access either `element.ref` or `element.props.ref` based on it. The version check seems to be faster than `propertyIsEnumerable` and IMO it communicates intent better.

Reply:
```
export default function getReactElementRef(element: React.ReactElement): React.Ref<any> | null {
  const majorVersion = parseInt(React.version.split('.')[0], 10);

  // 'ref' is passed as prop in React 19, whereas 'ref' is directly attached to children in older versions
  if (majorVersion >= 19) {
    return element.props.ref;
  }
  // @ts-expect-error element.ref is not included in the ReactElement type
  // https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/70189
  return element.ref;
}
```